File: src/core/validation.ts
import { validateWord } from "./phonology";
import { flattenLexicon, loadSpec, readSpecYaml } from "./spec";
import type { CompoundsSpec, CorpusSpec, DomainsSpec, EthraSpec, CorpusPlanSpec } from "./types";

export interface ValidationIssue {
  code: string;
  message: string;
  severity: "error" | "warning";
}

export interface SpecValidationReport {
  valid: boolean;
  issueCount: number;
  errors: ValidationIssue[];
  warnings: ValidationIssue[];
  stats: {
    roots: number;
    lexiconEntries: number;
    derivationPatterns: number;
    particles: number;
    pronouns: number;
    examples: number;
  };
}

export interface CorpusValidationReport {
  valid: boolean;
  issueCount: number;
  errors: ValidationIssue[];
  warnings: ValidationIssue[];
  stats: {
    items: number;
    tracks: number;
    domains: number;
    uniqueTerms: number;
  };
}

export interface CompoundsValidationReport {
  valid: boolean;
  issueCount: number;
  errors: ValidationIssue[];
  warnings: ValidationIssue[];
  stats: {
    terms: number;
    accepted: number;
    domains: number;
    registers: number;
  };
}

const allowedReservedOverlaps = new Map<string, string>([
  ["MR.object", "mor is grammaticalized as the inherited-duty particle."],
  ["VL.noun", "vel is grammaticalized as the choice particle."],
  ["DV.object", "dov is grammaticalized as the vow particle."],
  ["TR.object", "tor is grammaticalized as the chosen-duty particle."],
  ["HN.object", "hon is grammaticalized as the household collective pronoun."],
  ["SL.object", "sol is grammaticalized as the truth-intensifying particle."]
]);

function addIssue(issues: ValidationIssue[], severity: ValidationIssue["severity"], code: string, message: string): void {
  issues.push({ severity, code, message });
}

export function validateSpec(spec: EthraSpec = loadSpec()): SpecValidationReport {
  const issues: ValidationIssue[] = [];
  const rootIds = new Map<string, string>();
  const aliases = new Map<string, string>();
  const derivedWords = new Map<string, string>();
  const reservedWords = new Map<string, string>();
  const patternIds = new Set<string>();

  for (const pattern of spec.derivation_patterns.patterns) {
    if (patternIds.has(pattern.id)) {
      addIssue(issues, "error", "duplicate-derivation-pattern", `Derivation pattern '${pattern.id}' appears more than once.`);
    }
    patternIds.add(pattern.id);

    const validation = validateWord(pattern.example_word);
    if (!validation.valid) {
      addIssue(
        issues,
        "error",
        "invalid-pattern-example",
        `Derivation pattern '${pattern.id}' example '${pattern.example_word}' fails phonology: ${validation.errors.join("; ")}.`
      );
    }
  }

  if (spec.derivation_patterns.total_patterns_per_root !== spec.derivation_patterns.patterns.length) {
    addIssue(
      issues,
      "warning",
      "pattern-count-mismatch",
      `Derivation pattern count ${spec.derivation_patterns.total_patterns_per_root} does not match ${spec.derivation_patterns.patterns.length} listed patterns.`
    );
  }

  for (const particle of spec.particles.particles) {
    reservedWords.set(particle.word, `particle:${particle.type}`);
  }

  for (const pronoun of spec.pronouns.pronouns) {
    if (reservedWords.has(pronoun.word)) {
      addIssue(
        issues,
        "warning",
        "reserved-word-overlap",
        `Pronoun '${pronoun.word}' also appears as ${reservedWords.get(pronoun.word)}.`
      );
    }
    reservedWords.set(pronoun.word, `pronoun:${pronoun.stance}`);
  }

  for (const root of spec.roots.roots) {
    if (rootIds.has(root.id)) {
      addIssue(issues, "error", "duplicate-root-id", `Root '${root.id}' duplicates ${rootIds.get(root.id)}.`);
    }
    rootIds.set(root.id, root.semantic_field);

    for (const alias of root.aliases ?? []) {
      if (aliases.has(alias)) {
        addIssue(issues, "error", "duplicate-root-alias", `Alias '${alias}' is shared by ${aliases.get(alias)} and ${root.id}.`);
      }
      aliases.set(alias, root.id);
    }

    for (const [pattern, derived] of Object.entries(root.derived)) {
      if (!patternIds.has(pattern)) {
        addIssue(issues, "error", "unknown-root-pattern", `${root.id}.${pattern} is not listed in derivation-patterns.yaml.`);
      }

      const validation = validateWord(derived.word);
      if (!validation.valid) {
        addIssue(
          issues,
          "error",
          "invalid-derived-form",
          `${root.id}.${pattern} '${derived.word}' fails phonology: ${validation.errors.join("; ")}.`
        );
      }

      const existing = derivedWords.get(derived.word);
      if (existing) {
        addIssue(issues, "error", "duplicate-derived-word", `${root.id}.${pattern} '${derived.word}' duplicates ${existing}.`);
      }
      derivedWords.set(derived.word, `${root.id}.${pattern}`);

      const reserved = reservedWords.get(derived.word);
      if (reserved) {
        const overlapKey = `${root.id}.${pattern}`;
        const allowedReason = allowedReservedOverlaps.get(overlapKey);
        if (allowedReason) {
          addIssue(
            issues,
            "warning",
            "grammaticalized-overlap",
            `${overlapKey} '${derived.word}' overlaps with ${reserved}: ${allowedReason}`
          );
        } else {
          addIssue(issues, "error", "derived-reserved-collision", `${overlapKey} '${derived.word}' collides with ${reserved}.`);
        }
      }
    }

    for (const pattern of patternIds) {
      if (!root.derived[pattern]) {
        addIssue(issues, "error", "missing-root-pattern", `${root.id} is missing derivation pattern '${pattern}'.`);
      }
    }
  }

  const lexiconEntries = flattenLexicon(spec);
  for (const entry of lexiconEntries) {
    const validation = validateWord(entry.word);
    if (!validation.valid) {
      addIssue(
        issues,
        "warning",
        "lexicon-phonology",
        `Lexicon entry '${entry.word}' fails phonology: ${validation.errors.join("; ")}.`
      );
    }
  }

  const errors = issues.filter((issue) => issue.severity === "error");
  const warnings = issues.filter((issue) => issue.severity === "warning");

  return {
    valid: errors.length === 0,
    issueCount: issues.length,
    errors,
    warnings,
    stats: {
      roots: spec.roots.roots.length,
      lexiconEntries: lexiconEntries.length,
      derivationPatterns: spec.derivation_patterns.patterns.length,
      particles: spec.particles.particles.length,
      pronouns: spec.pronouns.pronouns.length,
      examples: spec.examples.examples.length
    }
  };
}

function normalizedTokenSet(spec: EthraSpec): Set<string> {
  const known = new Set<string>();

  for (const entry of flattenLexicon(spec)) {
    known.add(entry.word.toLowerCase());
  }
  for (const particle of spec.particles.particles) {
    known.add(particle.word.toLowerCase());
  }
  for (const pronoun of spec.pronouns.pronouns) {
    known.add(pronoun.word.toLowerCase());
  }
  for (const root of spec.roots.roots) {
    for (const derived of Object.values(root.derived)) {
      known.add(derived.word.toLowerCase());
    }
  }

  return known;
}

function tokenizeEthra(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[.,!?;:()"']/g, " ")
    .split(/\s+/)
    .map((token) => token.trim())
    .filter(Boolean);
}

function isKnownToken(token: string, known: Set<string>): boolean {
  if (known.has(token)) return true;
  if (token.includes("-")) {
    return token.split("-").every((part) => known.has(part));
  }
  return false;
}

export function validateCorpus(
  corpus: CorpusSpec = readSpecYaml<CorpusSpec>("corpus.yaml"),
  spec: EthraSpec = loadSpec()
): CorpusValidationReport {
  const issues: ValidationIssue[] = [];
  const known = normalizedTokenSet(spec);
  const plan = readSpecYaml<CorpusPlanSpec>("corpus-plan.yaml");
  const domains = readSpecYaml<DomainsSpec>("domains.yaml");
  const validTracks = new Set(plan.corpus_tracks.map((track) => track.id));
  const validDomains = new Set(domains.domains.map((domain) => domain.id));
  const ids = new Set<string>();
  const uniqueTerms = new Set<string>();

  for (const item of corpus.items) {
    if (ids.has(item.id)) {
      addIssue(issues, "error", "duplicate-corpus-id", `Corpus item '${item.id}' appears more than once.`);
    }
    ids.add(item.id);

    if (!validTracks.has(item.track)) {
      addIssue(issues, "error", "unknown-corpus-track", `${item.id} uses unknown track '${item.track}'.`);
    }

    for (const domain of item.domain_tags) {
      if (!validDomains.has(domain)) {
        addIssue(issues, "error", "unknown-corpus-domain", `${item.id} uses unknown domain tag '${domain}'.`);
      }
    }

    for (const field of ["english", "ethra", "literal", "notes", "register"] as const) {
      if (!item[field]?.trim()) {
        addIssue(issues, "error", "missing-corpus-field", `${item.id} is missing '${field}'.`);
      }
    }

    const tokens = tokenizeEthra(item.ethra);
    for (const token of tokens) {
      if (!isKnownToken(token, known)) {
        addIssue(issues, "error", "unknown-corpus-token", `${item.id} uses unknown Ethra token '${token}'.`);
      }
    }

    for (const term of item.terms) {
      const normalized = term.toLowerCase();
      uniqueTerms.add(normalized);
      if (!isKnownToken(normalized, known)) {
        addIssue(issues, "error", "unknown-corpus-term", `${item.id} lists unknown term '${term}'.`);
      }
    }
  }

  const trackCount = new Set(corpus.items.map((item) => item.track)).size;
  const domainCount = new Set(corpus.items.flatMap((item) => item.domain_tags)).size;
  const errors = issues.filter((issue) => issue.severity === "error");
  const warnings = issues.filter((issue) => issue.severity === "warning");

  return {
    valid: errors.length === 0,
    issueCount: issues.length,
    errors,
    warnings,
    stats: {
      items: corpus.items.length,
      tracks: trackCount,
      domains: domainCount,
      uniqueTerms: uniqueTerms.size
    }
  };
}

export function validateCompounds(
  compounds: CompoundsSpec = readSpecYaml<CompoundsSpec>("compounds.yaml"),
  spec: EthraSpec = loadSpec()
): CompoundsValidationReport {
  const issues: ValidationIssue[] = [];
  const known = normalizedTokenSet(spec);
  const domains = readSpecYaml<DomainsSpec>("domains.yaml");
  const validDomains = new Set(domains.domains.map((domain) => domain.id));
  const validStatuses = new Set(["candidate", "provisional", "accepted", "deprecated", "historical"]);
  const ids = new Set<string>();
  const words = new Set<string>();
  const registers = new Set<string>();
  const usedDomains = new Set<string>();

  for (const term of compounds.terms) {
    known.add(term.word.toLowerCase());
  }

  for (const term of compounds.terms) {
    const word = term.word.toLowerCase();
    registers.add(term.register);

    if (ids.has(term.id)) {
      addIssue(issues, "error", "duplicate-compound-id", `Compound '${term.id}' appears more than once.`);
    }
    ids.add(term.id);

    if (words.has(word)) {
      addIssue(issues, "error", "duplicate-compound-word", `Compound word '${term.word}' appears more than once.`);
    }
    words.add(word);

    if (!validStatuses.has(term.status)) {
      addIssue(issues, "error", "invalid-compound-status", `${term.id} has invalid status '${term.status}'.`);
    }

    if (term.components.length < 2) {
      addIssue(issues, "error", "compound-too-short", `${term.id} must have at least two components.`);
    }

    const expectedWord = term.components.join("-").toLowerCase();
    if (word !== expectedWord) {
      addIssue(issues, "error", "compound-word-mismatch", `${term.id} word '${term.word}' should be '${expectedWord}' from components.`);
    }

    if (!term.components.includes(term.head)) {
      addIssue(issues, "error", "compound-head-missing", `${term.id} head '${term.head}' is not one of its components.`);
    }

    const phonology = validateWord(term.word);
    if (!phonology.valid) {
      addIssue(issues, "error", "compound-phonology", `${term.id} '${term.word}' fails phonology: ${phonology.errors.join("; ")}.`);
    }

    for (const domain of term.domain_tags) {
      usedDomains.add(domain);
      if (!validDomains.has(domain)) {
        addIssue(issues, "error", "unknown-compound-domain", `${term.id} uses unknown domain tag '${domain}'.`);
      }
    }

    for (const component of term.components) {
      if (!isKnownToken(component.toLowerCase(), known)) {
        addIssue(issues, "error", "unknown-compound-component", `${term.id} uses unknown component '${component}'.`);
      }
    }

    for (const token of tokenizeEthra(term.example)) {
      if (!isKnownToken(token, known)) {
        addIssue(issues, "error", "unknown-compound-example-token", `${term.id} example uses unknown token '${token}'.`);
      }
    }
  }

  const errors = issues.filter((issue) => issue.severity === "error");
  const warnings = issues.filter((issue) => issue.severity === "warning");

  return {
    valid: errors.length === 0,
    issueCount: issues.length,
    errors,
    warnings,
    stats: {
      terms: compounds.terms.length,
      accepted: compounds.terms.filter((term) => term.status === "accepted").length,
      domains: usedDomains.size,
      registers: registers.size
    }
  };
}

File: src/validation.rs
use crate::phonology::validate_word;
use crate::spec::{flatten_lexicon, load_spec, read_spec_yaml};
use crate::types::{CompoundsSpec, CorpusPlanSpec, CorpusSpec, DomainsSpec, EthraSpec};
use anyhow::Result;
use serde::Serialize;
use std::collections::{HashMap, HashSet};

#[derive(Debug, Clone, Serialize)]
pub struct ValidationIssue {
    pub code: String,
    pub message: String,
    pub severity: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct SpecValidationStats {
    pub roots: usize,
    #[serde(rename = "lexiconEntries")]
    pub lexicon_entries: usize,
    #[serde(rename = "derivationPatterns")]
    pub derivation_patterns: usize,
    pub particles: usize,
    pub pronouns: usize,
    pub examples: usize,
}

#[derive(Debug, Clone, Serialize)]
pub struct SpecValidationReport {
    pub valid: bool,
    #[serde(rename = "issueCount")]
    pub issue_count: usize,
    pub errors: Vec<ValidationIssue>,
    pub warnings: Vec<ValidationIssue>,
    pub stats: SpecValidationStats,
}

#[derive(Debug, Clone, Serialize)]
pub struct CorpusValidationStats {
    pub items: usize,
    pub tracks: usize,
    pub domains: usize,
    #[serde(rename = "uniqueTerms")]
    pub unique_terms: usize,
}

#[derive(Debug, Clone, Serialize)]
pub struct CorpusValidationReport {
    pub valid: bool,
    #[serde(rename = "issueCount")]
    pub issue_count: usize,
    pub errors: Vec<ValidationIssue>,
    pub warnings: Vec<ValidationIssue>,
    pub stats: CorpusValidationStats,
}

#[derive(Debug, Clone, Serialize)]
pub struct CompoundsValidationStats {
    pub terms: usize,
    pub accepted: usize,
    pub domains: usize,
    pub registers: usize,
}

#[derive(Debug, Clone, Serialize)]
pub struct CompoundsValidationReport {
    pub valid: bool,
    #[serde(rename = "issueCount")]
    pub issue_count: usize,
    pub errors: Vec<ValidationIssue>,
    pub warnings: Vec<ValidationIssue>,
    pub stats: CompoundsValidationStats,
}

fn add_issue(
    issues: &mut Vec<ValidationIssue>,
    severity: &str,
    code: &str,
    message: impl Into<String>,
) {
    issues.push(ValidationIssue {
        severity: severity.to_string(),
        code: code.to_string(),
        message: message.into(),
    });
}

fn allowed_reserved_overlap(key: &str) -> Option<&'static str> {
    match key {
        "MR.object" => Some("mor is grammaticalized as the inherited-duty particle."),
        "VL.noun" => Some("vel is grammaticalized as the choice particle."),
        "DV.object" => Some("dov is grammaticalized as the vow particle."),
        "TR.object" => Some("tor is grammaticalized as the chosen-duty particle."),
        "HN.object" => Some("hon is grammaticalized as the household collective pronoun."),
        "SL.object" => Some("sol is grammaticalized as the truth-intensifying particle."),
        _ => None,
    }
}

pub fn validate_spec(spec: &EthraSpec) -> SpecValidationReport {
    let mut issues = Vec::new();
    let mut root_ids: HashMap<String, String> = HashMap::new();
    let mut aliases: HashMap<String, String> = HashMap::new();
    let mut derived_words: HashMap<String, String> = HashMap::new();
    let mut reserved_words: HashMap<String, String> = HashMap::new();
    let mut pattern_ids = HashSet::new();

    for pattern in &spec.derivation_patterns.patterns {
        if pattern_ids.contains(&pattern.id) {
            add_issue(
                &mut issues,
                "error",
                "duplicate-derivation-pattern",
                format!(
                    "Derivation pattern '{}' appears more than once.",
                    pattern.id
                ),
            );
        }
        pattern_ids.insert(pattern.id.clone());

        let validation = validate_word(&pattern.example_word);
        if !validation.valid {
            add_issue(
                &mut issues,
                "error",
                "invalid-pattern-example",
                format!(
                    "Derivation pattern '{}' example '{}' fails phonology: {}.",
                    pattern.id,
                    pattern.example_word,
                    validation.errors.join("; ")
                ),
            );
        }
    }

    if spec.derivation_patterns.total_patterns_per_root != spec.derivation_patterns.patterns.len() {
        add_issue(
            &mut issues,
            "warning",
            "pattern-count-mismatch",
            format!(
                "Derivation pattern count {} does not match {} listed patterns.",
                spec.derivation_patterns.total_patterns_per_root,
                spec.derivation_patterns.patterns.len()
            ),
        );
    }

    for particle in &spec.particles.particles {
        reserved_words.insert(particle.word.clone(), format!("particle:{}", particle.kind));
    }

    for pronoun in &spec.pronouns.pronouns {
        if let Some(existing) = reserved_words.get(&pronoun.word) {
            add_issue(
                &mut issues,
                "warning",
                "reserved-word-overlap",
                format!("Pronoun '{}' also appears as {}.", pronoun.word, existing),
            );
        }
        reserved_words.insert(pronoun.word.clone(), format!("pronoun:{}", pronoun.stance));
    }

    for root in &spec.roots.roots {
        if let Some(existing) = root_ids.get(&root.id) {
            add_issue(
                &mut issues,
                "error",
                "duplicate-root-id",
                format!("Root '{}' duplicates {}.", root.id, existing),
            );
        }
        root_ids.insert(root.id.clone(), root.semantic_field.clone());

        for alias in &root.aliases {
            if let Some(existing) = aliases.get(alias) {
                add_issue(
                    &mut issues,
                    "error",
                    "duplicate-root-alias",
                    format!(
                        "Alias '{}' is shared by {} and {}.",
                        alias, existing, root.id
                    ),
                );
            }
            aliases.insert(alias.clone(), root.id.clone());
        }

        for (pattern, derived) in &root.derived {
            if !pattern_ids.contains(pattern) {
                add_issue(
                    &mut issues,
                    "error",
                    "unknown-root-pattern",
                    format!(
                        "{}.{} is not listed in derivation-patterns.yaml.",
                        root.id, pattern
                    ),
                );
            }

            let validation = validate_word(&derived.word);
            if !validation.valid {
                add_issue(
                    &mut issues,
                    "error",
                    "invalid-derived-form",
                    format!(
                        "{}.{} '{}' fails phonology: {}.",
                        root.id,
                        pattern,
                        derived.word,
                        validation.errors.join("; ")
                    ),
                );
            }

            if let Some(existing) = derived_words.get(&derived.word) {
                add_issue(
                    &mut issues,
                    "error",
                    "duplicate-derived-word",
                    format!(
                        "{}.{} '{}' duplicates {}.",
                        root.id, pattern, derived.word, existing
                    ),
                );
            }
            derived_words.insert(derived.word.clone(), format!("{}.{}", root.id, pattern));

            if let Some(reserved) = reserved_words.get(&derived.word) {
                let overlap_key = format!("{}.{}", root.id, pattern);
                if let Some(reason) = allowed_reserved_overlap(&overlap_key) {
                    add_issue(
                        &mut issues,
                        "warning",
                        "grammaticalized-overlap",
                        format!(
                            "{} '{}' overlaps with {}: {}",
                            overlap_key, derived.word, reserved, reason
                        ),
                    );
                } else {
                    add_issue(
                        &mut issues,
                        "error",
                        "derived-reserved-collision",
                        format!(
                            "{} '{}' collides with {}.",
                            overlap_key, derived.word, reserved
                        ),
                    );
                }
            }
        }

        for pattern in &pattern_ids {
            if !root.derived.contains_key(pattern) {
                add_issue(
                    &mut issues,
                    "error",
                    "missing-root-pattern",
                    format!("{} is missing derivation pattern '{}'.", root.id, pattern),
                );
            }
        }
    }

    let lexicon_entries = flatten_lexicon(spec);
    for entry in &lexicon_entries {
        let validation = validate_word(&entry.word);
        if !validation.valid {
            add_issue(
                &mut issues,
                "warning",
                "lexicon-phonology",
                format!(
                    "Lexicon entry '{}' fails phonology: {}.",
                    entry.word,
                    validation.errors.join("; ")
                ),
            );
        }
    }

    let known = normalized_token_set(spec);
    for example in &spec.examples.examples {
        for token in tokenize_ethra(&example.ethra) {
            if !is_known_token(&token, &known) {
                add_issue(
                    &mut issues,
                    "error",
                    "unknown-example-token",
                    format!(
                        "Canonical example {} uses unknown Ethra token '{}'.",
                        example.number, token
                    ),
                );
            }
        }
    }

    let errors = issues
        .iter()
        .filter(|issue| issue.severity == "error")
        .cloned()
        .collect::<Vec<_>>();
    let warnings = issues
        .iter()
        .filter(|issue| issue.severity == "warning")
        .cloned()
        .collect::<Vec<_>>();
    SpecValidationReport {
        valid: errors.is_empty(),
        issue_count: issues.len(),
        errors,
        warnings,
        stats: SpecValidationStats {
            roots: spec.roots.roots.len(),
            lexicon_entries: lexicon_entries.len(),
            derivation_patterns: spec.derivation_patterns.patterns.len(),
            particles: spec.particles.particles.len(),
            pronouns: spec.pronouns.pronouns.len(),
            examples: spec.examples.examples.len(),
        },
    }
}

fn normalized_token_set(spec: &EthraSpec) -> HashSet<String> {
    let mut known = HashSet::new();
    for entry in flatten_lexicon(spec) {
        known.insert(entry.word.to_lowercase());
    }
    for particle in &spec.particles.particles {
        known.insert(particle.word.to_lowercase());
    }
    for pronoun in &spec.pronouns.pronouns {
        known.insert(pronoun.word.to_lowercase());
    }
    for root in &spec.roots.roots {
        for derived in root.derived.values() {
            known.insert(derived.word.to_lowercase());
        }
    }
    known
}

fn tokenize_ethra(text: &str) -> Vec<String> {
    text.to_lowercase()
        .chars()
        .map(|char_value| {
            if matches!(
                char_value,
                '.' | ',' | '!' | '?' | ';' | ':' | '(' | ')' | '"' | '\''
            ) {
                ' '
            } else {
                char_value
            }
        })
        .collect::<String>()
        .split_whitespace()
        .map(str::trim)
        .filter(|token| !token.is_empty())
        .map(ToOwned::to_owned)
        .collect()
}

fn is_known_token(token: &str, known: &HashSet<String>) -> bool {
    known.contains(token)
        || (token.contains('-') && token.split('-').all(|part| known.contains(part)))
}

pub fn validate_corpus(corpus: CorpusSpec, spec: &EthraSpec) -> Result<CorpusValidationReport> {
    let mut issues = Vec::new();
    let known = normalized_token_set(spec);
    let plan = read_spec_yaml::<CorpusPlanSpec>("corpus-plan.yaml")?;
    let domains = read_spec_yaml::<DomainsSpec>("domains.yaml")?;
    let valid_tracks = plan
        .corpus_tracks
        .into_iter()
        .map(|track| track.id)
        .collect::<HashSet<_>>();
    let valid_domains = domains
        .domains
        .into_iter()
        .map(|domain| domain.id)
        .collect::<HashSet<_>>();
    let mut ids = HashSet::new();
    let mut unique_terms = HashSet::new();

    for item in &corpus.items {
        if ids.contains(&item.id) {
            add_issue(
                &mut issues,
                "error",
                "duplicate-corpus-id",
                format!("Corpus item '{}' appears more than once.", item.id),
            );
        }
        ids.insert(item.id.clone());

        if !valid_tracks.contains(&item.track) {
            add_issue(
                &mut issues,
                "error",
                "unknown-corpus-track",
                format!("{} uses unknown track '{}'.", item.id, item.track),
            );
        }

        for domain in &item.domain_tags {
            if !valid_domains.contains(domain) {
                add_issue(
                    &mut issues,
                    "error",
                    "unknown-corpus-domain",
                    format!("{} uses unknown domain tag '{}'.", item.id, domain),
                );
            }
        }

        if item.english.trim().is_empty() {
            add_issue(
                &mut issues,
                "error",
                "missing-corpus-field",
                format!("{} is missing 'english'.", item.id),
            );
        }
        if item.ethra.trim().is_empty() {
            add_issue(
                &mut issues,
                "error",
                "missing-corpus-field",
                format!("{} is missing 'ethra'.", item.id),
            );
        }
        if item.literal.trim().is_empty() {
            add_issue(
                &mut issues,
                "error",
                "missing-corpus-field",
                format!("{} is missing 'literal'.", item.id),
            );
        }
        if item.notes.trim().is_empty() {
            add_issue(
                &mut issues,
                "error",
                "missing-corpus-field",
                format!("{} is missing 'notes'.", item.id),
            );
        }
        if item.register.trim().is_empty() {
            add_issue(
                &mut issues,
                "error",
                "missing-corpus-field",
                format!("{} is missing 'register'.", item.id),
            );
        }

        for token in tokenize_ethra(&item.ethra) {
            if !is_known_token(&token, &known) {
                add_issue(
                    &mut issues,
                    "error",
                    "unknown-corpus-token",
                    format!("{} uses unknown Ethra token '{}'.", item.id, token),
                );
            }
        }

        for term in &item.terms {
            let normalized = term.to_lowercase();
            unique_terms.insert(normalized.clone());
            if !is_known_token(&normalized, &known) {
                add_issue(
                    &mut issues,
                    "error",
                    "unknown-corpus-term",
                    format!("{} lists unknown term '{}'.", item.id, term),
                );
            }
        }
    }

    let track_count = corpus
        .items
        .iter()
        .map(|item| item.track.clone())
        .collect::<HashSet<_>>()
        .len();
    let domain_count = corpus
        .items
        .iter()
        .flat_map(|item| item.domain_tags.clone())
        .collect::<HashSet<_>>()
        .len();
    let errors = issues
        .iter()
        .filter(|issue| issue.severity == "error")
        .cloned()
        .collect::<Vec<_>>();
    let warnings = issues
        .iter()
        .filter(|issue| issue.severity == "warning")
        .cloned()
        .collect::<Vec<_>>();
    Ok(CorpusValidationReport {
        valid: errors.is_empty(),
        issue_count: issues.len(),
        errors,
        warnings,
        stats: CorpusValidationStats {
            items: corpus.items.len(),
            tracks: track_count,
            domains: domain_count,
            unique_terms: unique_terms.len(),
        },
    })
}

pub fn validate_compounds(
    compounds: CompoundsSpec,
    spec: &EthraSpec,
) -> Result<CompoundsValidationReport> {
    let mut issues = Vec::new();
    let mut known = normalized_token_set(spec);
    let domains = read_spec_yaml::<DomainsSpec>("domains.yaml")?;
    let valid_domains = domains
        .domains
        .into_iter()
        .map(|domain| domain.id)
        .collect::<HashSet<_>>();
    let valid_statuses = HashSet::from([
        "candidate".to_string(),
        "provisional".to_string(),
        "accepted".to_string(),
        "deprecated".to_string(),
        "historical".to_string(),
    ]);
    let mut ids = HashSet::new();
    let mut words = HashSet::new();
    let mut registers = HashSet::new();
    let mut used_domains = HashSet::new();

    for term in &compounds.terms {
        known.insert(term.word.to_lowercase());
    }

    for term in &compounds.terms {
        let word = term.word.to_lowercase();
        registers.insert(term.register.clone());

        if ids.contains(&term.id) {
            add_issue(
                &mut issues,
                "error",
                "duplicate-compound-id",
                format!("Compound '{}' appears more than once.", term.id),
            );
        }
        ids.insert(term.id.clone());

        if words.contains(&word) {
            add_issue(
                &mut issues,
                "error",
                "duplicate-compound-word",
                format!("Compound word '{}' appears more than once.", term.word),
            );
        }
        words.insert(word.clone());

        if !valid_statuses.contains(&term.status.to_string()) {
            add_issue(
                &mut issues,
                "error",
                "invalid-compound-status",
                format!("{} has invalid status '{}'.", term.id, term.status),
            );
        }

        if term.components.len() < 2 {
            add_issue(
                &mut issues,
                "error",
                "compound-too-short",
                format!("{} must have at least two components.", term.id),
            );
        }

        let expected_word = term.components.join("-").to_lowercase();
        if word != expected_word {
            add_issue(
                &mut issues,
                "error",
                "compound-word-mismatch",
                format!(
                    "{} word '{}' should be '{}' from components.",
                    term.id, term.word, expected_word
                ),
            );
        }

        if !term.components.contains(&term.head) {
            add_issue(
                &mut issues,
                "error",
                "compound-head-missing",
                format!(
                    "{} head '{}' is not one of its components.",
                    term.id, term.head
                ),
            );
        }

        let phonology = validate_word(&term.word);
        if !phonology.valid {
            add_issue(
                &mut issues,
                "error",
                "compound-phonology",
                format!(
                    "{} '{}' fails phonology: {}.",
                    term.id,
                    term.word,
                    phonology.errors.join("; ")
                ),
            );
        }

        for domain in &term.domain_tags {
            used_domains.insert(domain.clone());
            if !valid_domains.contains(domain) {
                add_issue(
                    &mut issues,
                    "error",
                    "unknown-compound-domain",
                    format!("{} uses unknown domain tag '{}'.", term.id, domain),
                );
            }
        }

        for component in &term.components {
            if !is_known_token(&component.to_lowercase(), &known) {
                add_issue(
                    &mut issues,
                    "error",
                    "unknown-compound-component",
                    format!("{} uses unknown component '{}'.", term.id, component),
                );
            }
        }

        for token in tokenize_ethra(&term.example) {
            if !is_known_token(&token, &known) {
                add_issue(
                    &mut issues,
                    "error",
                    "unknown-compound-example-token",
                    format!("{} example uses unknown token '{}'.", term.id, token),
                );
            }
        }
    }

    let errors = issues
        .iter()
        .filter(|issue| issue.severity == "error")
        .cloned()
        .collect::<Vec<_>>();
    let warnings = issues
        .iter()
        .filter(|issue| issue.severity == "warning")
        .cloned()
        .collect::<Vec<_>>();
    Ok(CompoundsValidationReport {
        valid: errors.is_empty(),
        issue_count: issues.len(),
        errors,
        warnings,
        stats: CompoundsValidationStats {
            terms: compounds.terms.len(),
            accepted: compounds
                .terms
                .iter()
                .filter(|term| term.status.to_string() == "accepted")
                .count(),
            domains: used_domains.len(),
            registers: registers.len(),
        },
    })
}

pub fn validate_current_spec() -> Result<SpecValidationReport> {
    Ok(validate_spec(load_spec()?))
}

pub fn validate_current_corpus() -> Result<CorpusValidationReport> {
    validate_corpus(read_spec_yaml("corpus.yaml")?, load_spec()?)
}

pub fn validate_current_compounds() -> Result<CompoundsValidationReport> {
    validate_compounds(read_spec_yaml("compounds.yaml")?, load_spec()?)
}
